Test: cts: Reimplement cts-regression in Python

nrwahl2 opened the pull request:
Closes T498

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,257 @@
+    if os.access(cmd[0], os.X_OK):
+        # Test might not be executable if run from source directory
+        mode = os.stat(cmd[0]).st_mode
+        os.chmod(cmd[0], mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
```
It's not at all clear to me that this is needed or desirable. However, `chmod a+x` was in the bash version, and this is the Python equivalent.

kgaillot replied:
Now that the individual tests are all .in files, and properly configured in ./configure to be executable, it's not needed. There was a time when that wasn't true for all tests and this was useful.

nrwahl2 replied:
Gotcha, removed.

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,257 @@
+    Replace the 'all' alias with the tests that it represents.
+    Get rid of duplicates.
+
+    The main motivation is that when the `choices` argument of
```
Just trust me on this haha... it was a pain and this is the cleanest thing I could come up with after consulting with the docs and StackOverflow. This type of thing is more or less the intended use case of subclassing `argparse.Action`.

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,257 @@
+class CrmExit(IntEnum):
+    """Pacemaker exit codes.
+
+    These values must be kept in sync with include/crm/common/results.h.
```
Incidentally the ones in `cts-fencing` haven't been. The comment there even refers to an old header file.

kgaillot replied:
In sync just means the values must match for any that are actually used -- we don't need to define constants that aren't used, and the names aren't essential. However once the constants are moved to a module, listing them all will be more useful, so this is worth keeping here for now.

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,257 @@
+    ''')
+    if REMOTE_ENABLED:
+        description += (
+            ' pacemaker_remote  Resource agent executor in remote mode'
```
This can be moved underneath `exec` if we really want to, with `fencing` and `all` underneath `pacemaker_remote`. This reordering made the string formatting a lot cleaner and I figure it doesn't matter much.

kgaillot replied:
The description can be a component class member as well, and this reduces nicely

nrwahl2 replied:
Done.

kgaillot commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,257 @@
+    if test == 'pacemaker_remote':
+        cmd = [os.path.join(test_home, 'cts-exec'), '-R']
+    else:
+        cmd = [os.path.join(test_home, 'cts-%s' % test)]
```
Looking ahead to when all the tests will eventually be modularized, I think it will be worthwhile to define a regression test component class. The constructor can define this command as a member or method return.

The hope for the eventual design is that they can all share infrastructure code and the only real difference between them is a list of specific tests. I'm thinking something like an individual test class, a test group of some sort (equivalent to cts-cli's dates tools crm_mon etc.), and a component class (equivalent to cli scheduler etc.).

So we'd be starting at the top with a component class.

nrwahl2 replied:
Done (`Component` only so far).

kgaillot commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,257 @@
+
+    # Valgrind is supported only for cli and scheduler currently
+    if valgrind and test in ['cli', 'scheduler']:
+        cmd.append('--valgrind')
```
Obviously this check could be a component class boolean member

nrwahl2 replied:
Done

kgaillot commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,257 @@
+        # Run as root
+        if os.geteuid() != 0:
+            print('Enter the sudo password if prompted')
+            cmd = ['sudo'] + cmd
```
Note cli/scheduler components don't need sudo. A component class boolean could control this as well

nrwahl2 replied:
Done

kgaillot reviewed (commented):
Feels like progress :)

nrwahl2 commented:
Replaced commit to address feedback. The main change is the addition of a `Component` class.

I also removed the explicit check for executability -- we'll rely on try/except. And I removed the python equivalent of `chmod a+x` -- as discussed in the review, that's no longer needed.

I'm doing a little bit more testing (should be done in 30 mins or so) but I'm hoping not to touch this, except in response to any other feedback.

nrwahl2 commented:
I noticed the mode changed from `100755 → 100644` according to the GitHub comparison, since I started with a new file and kept the default permissions. It builds just fine without executable perms on the `.in` file of course. Can I leave it as 644?

kgaillot commented:
> I noticed the mode changed from `100755 → 100644` according to the GitHub comparison, since I started with a new file and kept the default permissions. It builds just fine without executable perms on the `.in` file of course. Can I leave it as 644?

Yes, that's what .in files should be

nrwahl2 commented:
> I'm doing a little bit more testing (should be done in 30 mins or so) but I'm hoping not to touch this, except in response to any other feedback.

It's fine so far and I've done everything that comes to mind. Planning to be hands-off now.

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,335 @@
+
+
+REMOTE_ENABLED = bool('@PC_NAME_GNUTLS@')
+TEST_HOME = ''
```
Not sure this should be a "constant" (all caps, no true constants) in hindsight, since we set it in `main()`. Doesn't even have to be global. We could pass it as a constructor argument to `Component`. Feels redundant that way but perhaps it's more semantically correct.

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,335 @@
+
+    :param components: A list of names of components for which to run
+        tests.
+    :type tests: list[str]
```
Need to fix this docstring now that we're taking a list of `Component` objects and not a list of component names.

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,335 @@
+        TEST_HOME = os.path.dirname(sys.argv[0])
+
+    # Available components
+    components = {
```
Could optionally make a helper function (only called once, with hard-coded values plus the remote check) to construct this dict, so that `main()` is shorter.

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,335 @@
+    description = textwrap.dedent('''\
+        Run Pacemaker regression tests.
+
+        Available components (default components are 'scheduler cli'):
```
note to self: might as well flip order to 'cli scheduler' now, since program description output is sorted

nrwahl2 commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,335 @@
+    parser.add_argument('-v', '--valgrind', action='store_true',
+                        help='Run test commands under valgrind')
+    parser.add_argument('components', nargs='*', choices=choices,
+                        default='default',
```
By the way: I went with `'default'` because we need SOME kind of explicit default value if we use `choices`. We don't have to advertise the literal `'default'` in our description. But for an invalid argument, `argparse` will print the default value as one of the choices. Other options are `''` and `[]`, both of which feel misleading.
```
# with default='default' and 'default' in choices
$ cts/cts-regression asdf
usage: cts-regression [-h] [-V] [-v] [COMPONENT ...]
cts-regression: error: argument COMPONENT: invalid choice: 'asdf' (choose from 'cli', 'exec', 'fencing', 'pacemaker_remote', 'scheduler', 'all', 'default')
$ cts/cts-regression default
['cli', 'scheduler']

# with default=[] and [] in choices
# choices is no longer all strings, and error message incorrectly suggests user can pass [] from CLI
$ cts/cts-regression asdf
usage: cts-regression [-h] [-V] [-v] [COMPONENT ...]
cts-regression: error: argument COMPONENT: invalid choice: 'asdf' (choose from 'cli', 'exec', 'fencing', 'pacemaker_remote', 'scheduler', 'all', [])
$ cts/cts-regression []
usage: cts-regression [-h] [-V] [-v] [COMPONENT ...]
cts-regression: error: argument COMPONENT: invalid choice: '[]' (choose from 'cli', 'exec', 'fencing', 'pacemaker_remote', 'scheduler', 'all', [])

# with default='' and '' in choices
# error message incorrectly suggests user can pass '' from CLI
[reid@laptop pacemaker]$ cts/cts-regression asdf
usage: cts-regression [-h] [-V] [-v] [COMPONENT ...]
cts-regression: error: argument COMPONENT: invalid choice: 'asdf' (choose from 'cli', 'exec', 'fencing', 'pacemaker_remote', 'scheduler', 'all', '')
[reid@laptop pacemaker]$ cts/cts-regression ''
Traceback (most recent call last):

# All of the above work fine for "no argument on CLI"
```

It might be possible to get rid of `choices` altogether and handle everything in the custom action, but that's less obvious and would require hard-coding the choices into action class (can't pass constructor args manually AFAICT).

kgaillot replied:
'default' works for me, otherwise we could do something like 'essential', 'basic', or 'nonroot'

nrwahl2 commented:
Minor changes:
* Made `TEST_HOME` a constructor param for `Component`, instead of a global var.
* Renamed `Component.use_sudo` to `Component.requires_root`.
* Added `cmd` to `Component` class docstring.
* Corrected type of `components` arg in `run_components()` docstring.

kgaillot commented on cts/cts-regression.in:
```diff
@@ -1,230 +1,344 @@
+    PROMOTED = 8
+    FAILED_PROMOTED = 9
+
+    # sysexits.h USAGE = 64
```
Did you mean to have this on one line?

nrwahl2 replied:
nope

kgaillot reviewed (approved):
Looks good to me, let me know when you're ready to merge

nrwahl2 commented:
New push just moves `USAGE = 64` to its own line (https://github.com/ClusterLabs/pacemaker/pull/2806#discussion_r951460694).
